src/anemoi/models/layers/attention.py

```diff
@@ -89,24 +139,220 @@ def forward(
         value = shard_heads(value, shapes=shapes, mgroup=model_comm_group)
         dropout_p = self.dropout_p if self.training else 0.0
 
-        if _FLASH_ATTENTION_AVAILABLE:
-            query, key, value = (
-                einops.rearrange(t, "batch heads grid vars -> batch grid heads vars") for t in (query, key, value)
+        out = self.attention(
+            query,
+            key,
+            value,
+            batch_size,
+            causal=False,
+            window_size=self.window_size,
+            dropout_p=dropout_p,
+            softcap=self.softcap,
+            alibi_slopes=self.alibi_slopes,
+        )
+
+        out = shard_sequence(out, shapes=shapes, mgroup=model_comm_group)
+        out = einops.rearrange(out, "batch heads grid vars -> (batch grid) (heads vars)")
+
+        out = self.projection(out)
+
+        return out
+
+
+class TorchAttentionWrapper(nn.Module):
+    """Wrapper for Pytorch dot product attention"""
+
+    def __init__(self):
+        super().__init__()
+
+        from torch.nn.functional import scaled_dot_product_attention
+
+        self.attention = scaled_dot_product_attention
+        self.mask = None
+        self.window_size = None
+
+    def update_mask(self, seq_len, window_size: int, device: str):
+        update_mask = (
+            self.mask is None or self.window_size != window_size or tuple(self.mask.shape) != (seq_len, seq_len)
+        )
+        if update_mask:
+            self.window_size = window_size
+            self.mask = (
+                torch.abs(
+                    torch.arange(seq_len, device=device).unsqueeze(0)
+                    - torch.arange(seq_len, device=device).unsqueeze(1)
+                )
+                <= window_size
             )
-            out = self.attention(query, key, value, causal=False, window_size=self.window_size, dropout_p=dropout_p)
-            out = einops.rearrange(out, "batch grid heads vars -> batch heads grid vars")
+
+    def forward(
+        self,
+        query,
+        key,
+        value,
+        batch_size: int,
+        causal=False,
+        window_size=None,
+        dropout_p=0.0,
+        softcap=None,
+        alibi_slopes=None,
+    ):
+        if softcap is not None:
+            SystemError(
+                "Error. Softcap not supported by Pytorchs SDPA. please switch to flash attention or disable softcap."
+            )
+        if alibi_slopes is not None:
+            SystemError(
+                "Error. Alibi slopes not supported by Pytorchs SDPA. please switch to flash attention or disable alibi slopes."
+            )
+        if window_size is not None:
+            self.update_mask(query.shape[-2], window_size=window_size, device=query.device)
         else:
+            self.mask = None
+
+        with torch.nn.attention.sdpa_kernel(backends=[torch.nn.attention.SDPBackend.MATH]):
```

Comment:
What is the reasoning to limit the backend?

Reply:
@JesperDramsch the idea is to use the `MATH` backend because it comes with pytorch and doesn't rely on external libraries like the other two options:

* `FLASH_ATTENTION` which we allready have as an explicit option, or 
* `EFFICIENT_ATTENTION` [memory-efficient-attention](https://github.com/facebookresearch/xformers)

I would rather support the last one explicitly than switching via SDPA backend